Weird freeze and crash between zed, fish, alacritty

### Summary

<!-- Please insert a one line summary of the issue below -->
Zed freezes regularly with the message `fish: Job 1, 'zed .' has stopped` from fish

<!-- Include all steps necessary to reproduce from a clean Zed installation. Be verbose -->
Steps to trigger the problem:
1. Start zed from the command line in alacritty using fish shell
2. Use zed for a couple of minutes

Actual Behavior:

Zed freezes. When I switch focus Zed crashes. In the terminal I see `fish: Job 1, 'zed .' has stopped`. When I press enter, the terminal crashes as well.

Expected Behavior:

Not crash

### Zed Version and System Specs

Zed: v0.181.0 (Zed Dev 8b3eb98d8602e5f7d555f47add63aae11cb2939a) 
OS: Linux Wayland arch unknown
Memory: 31.1 GiB
Architecture: x86_64
GPU: Intel(R) Iris(R) Xe Graphics (ADL GT2) || Intel open-source Mesa driver || Mesa 25.0.2-arch1.2.1

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary>

<!-- Click below this line and paste or drag-and-drop your log-->
```
2025-01-06T19:32:55.32396264+01:00 [ERROR] oneshot canceled
2025-01-06T19:32:55.324061081+01:00 [ERROR] oneshot canceled
2025-01-06T19:32:55.32420304+01:00 [ERROR] oneshot canceled
2025-01-06T19:32:55.324281151+01:00 [ERROR] window not found
2025-01-06T19:32:55.32432779+01:00 [ERROR] window not found
2025-01-06T19:32:55.324360347+01:00 [ERROR] oneshot canceled
2025-01-06T19:32:55.324405936+01:00 [ERROR] window not found
2025-01-06T19:32:55.324477886+01:00 [ERROR] window not found
2025-01-06T19:32:55.324561025+01:00 [ERROR] oneshot canceled
```
<!-- Click above this line and paste or drag-and-drop your log--></details>

--- a/crates/project/src/environment.rs
+++ b/crates/project/src/environment.rs
@@ -273,23 +273,14 @@ async fn load_shell_environment(
     //
     // In certain shells we need to execute additional_command in order to
     // trigger the behavior of direnv, etc.
-    //
-    //
-    // The `exit 0` is the result of hours of debugging, trying to find out
-    // why running this command here, without `exit 0`, would mess
-    // up signal process for our process so that `ctrl-c` doesn't work
-    // anymore.
-    //
-    // We still don't know why `$SHELL -l -i -c '/usr/bin/env -0'`  would
-    // do that, but it does, and `exit 0` helps.
 
     let command = match shell_name {
         Some("fish") => format!(
-            "cd '{}'; emit fish_prompt; printf '%s' {MARKER}; /usr/bin/env; exit 0;",
+            "cd '{}'; emit fish_prompt; printf '%s' {MARKER}; /usr/bin/env;",
             dir.display()
         ),
         _ => format!(
-            "cd '{}'; printf '%s' {MARKER}; /usr/bin/env; exit 0;",
+            "cd '{}'; printf '%s' {MARKER}; /usr/bin/env;",
             dir.display()
         ),
     };
@@ -297,16 +288,21 @@ async fn load_shell_environment(
     // csh/tcsh only supports `-l` if it's the only flag. So this won't be a login shell.
     // Users must rely on vars from `~/.tcshrc` or `~/.cshrc` and not `.login` as a result.
     let args = match shell_name {
-        Some("tcsh") | Some("csh") => vec!["-i", "-c", &command],
-        _ => vec!["-l", "-i", "-c", &command],
+        Some("tcsh") | Some("csh") => vec!["-i".to_string(), "-c".to_string(), command],
+        _ => vec![
+            "-l".to_string(),
+            "-i".to_string(),
+            "-c".to_string(),
+            command,
+        ],
     };
 
-    let Some(output) = smol::process::Command::new(&shell)
-        .args(&args)
-        .output()
-        .await
-        .log_err()
-    else {
+    let Some(output) = smol::unblock(move || {
+        util::set_pre_exec_to_start_new_session(std::process::Command::new(&shell).args(&args))
+            .output()
+    })
+    .await
+    .log_err() else {
         return message(
             "Failed to spawn login shell to source login environment variables. See logs for details",
         );

--- a/crates/util/src/util.rs
+++ b/crates/util/src/util.rs
@@ -321,21 +321,16 @@ pub fn load_login_shell_environment() -> Result<()> {
         .and_then(|home| home.into_string().ok())
         .map(|home| format!("cd '{home}';"));
 
-    // The `exit 0` is the result of hours of debugging, trying to find out
-    // why running this command here, without `exit 0`, would mess
-    // up signal process for our process so that `ctrl-c` doesn't work
-    // anymore.
-    // We still don't know why `$SHELL -l -i -c '/usr/bin/env -0'`  would
-    // do that, but it does, and `exit 0` helps.
     let shell_cmd = format!(
-        "{}printf '%s' {marker}; /usr/bin/env; exit 0;",
+        "{}printf '%s' {marker}; /usr/bin/env;",
         shell_cmd_prefix.as_deref().unwrap_or("")
     );
 
-    let output = std::process::Command::new(&shell)
-        .args(["-l", "-i", "-c", &shell_cmd])
-        .output()
-        .context("failed to spawn login shell to source login environment variables")?;
+    let output = set_pre_exec_to_start_new_session(
+        std::process::Command::new(&shell).args(["-l", "-i", "-c", &shell_cmd]),
+    )
+    .output()
+    .context("failed to spawn login shell to source login environment variables")?;
     if !output.status.success() {
         Err(anyhow!("login shell exited with error"))?;
     }
@@ -355,6 +350,26 @@ pub fn load_login_shell_environment() -> Result<()> {
     }
 
     Ok(())
+}
+
+/// Configures the process to start a new session, to prevent interactive shells from taking control
+/// of the terminal.
+///
+/// For more details: https://registerspill.thorstenball.com/p/how-to-lose-control-of-your-shell
+pub fn set_pre_exec_to_start_new_session(
+    command: &mut std::process::Command,
+) -> &mut std::process::Command {
+    // safety: code in pre_exec should be signal safe.
+    // https://man7.org/linux/man-pages/man7/signal-safety.7.html
+    #[cfg(not(target_os = "windows"))]
+    unsafe {
+        use std::os::unix::process::CommandExt;
+        command.pre_exec(|| {
+            libc::setsid();
+            Ok(())
+        });
+    };
+    command
 }
 
 /// Parse the result of calling `usr/bin/env` with no arguments

--- a/crates/vim/Cargo.toml
+++ b/crates/vim/Cargo.toml
@@ -28,7 +28,6 @@ futures.workspace = true
 gpui.workspace = true
 itertools.workspace = true
 language.workspace = true
-libc.workspace = true
 log.workspace = true
 multi_buffer.workspace = true
 nvim-rs = { git = "https://github.com/KillTheMule/nvim-rs", branch = "master", features = ["use_tokio"], optional = true }

--- a/crates/vim/src/command.rs
+++ b/crates/vim/src/command.rs
@@ -1519,18 +1519,7 @@ impl ShellExec {
             process.stdin(Stdio::null());
         };
 
-        // https://registerspill.thorstenball.com/p/how-to-lose-control-of-your-shell
-        //
-        // safety: code in pre_exec should be signal safe.
-        // https://man7.org/linux/man-pages/man7/signal-safety.7.html
-        #[cfg(not(target_os = "windows"))]
-        unsafe {
-            use std::os::unix::process::CommandExt;
-            process.pre_exec(|| {
-                libc::setsid();
-                Ok(())
-            });
-        };
+        util::set_pre_exec_to_start_new_session(&mut process);
         let is_read = self.is_read;
 
         let task = cx.spawn_in(window, async move |vim, cx| {

